### PR TITLE
MLDB-2172 add a chunk recorder for the structured dataset

### PR DESCRIPTION
**Pull request by mathieumb**

### CODE REVIEW

**Comment by mathieumb** [comment]
Performance only.

File: builtin/sub_dataset.cc [hunk 1]
```diff
@@ -200,19 +292,31 @@ struct SubDataset::Itl
...
         
     }
 
+    void commitWrites(Chunk& newChunk) {
+        std::lock_guard<std::mutex> lock(recordLock);
+        //STACK_PROFILE(commitWrites);
+        size_t start = mainChunk.subOutput.size();
+        mainChunk.subOutput.insert( mainChunk.subOutput.begin(), make_move_iterator(newChunk.subOutput.begin()),  make_move_iterator(newChunk.subOutput.end()));
+        mainChunk.columnNames.insert( newChunk.columnNames.begin(),  newChunk.columnNames.end());
+        mainChunk.fullFlattenedColumnNames.insert( newChunk.fullFlattenedColumnNames.begin(), newChunk.fullFlattenedColumnNames.end());
```
Comment (FinchPowers) [inline builtin/sub_dataset.cc]: Lines too long

File: builtin/sub_dataset.cc [hunk 2]
```diff
@@ -122,43 +205,52 @@ struct SubDataset::Itl
...
 
     }
 
-    void AddRowInternal(const RowPath & rowName,
-                const ExpressionValue & expr)
+    template< typename E >
+    static void AddRowInternal(const RowPath & rowName,
+                        E & expr,
+                        Chunk& chunk, 
+                        bool destructive)
```
Comment (FinchPowers) [inline builtin/sub_dataset.cc]: Indent properly.

File: builtin/sub_dataset.cc [hunk 3]
```diff
@@ -281,62 +385,62 @@ struct SubDataset::Itl
...
     }
 
     virtual bool knownColumn(const ColumnPath & column) const
     {
         if (column.size() == 1)
-            return std::find(columnNames.begin(), columnNames.end(), column[0]) != columnNames.end();
+            return std::find(mainChunk.columnNames.begin(), mainChunk.columnNames.end(), column[0]) != mainChunk.columnNames.end();
         else
-            return std::find(fullFlattenedColumnNames.begin(), fullFlattenedColumnNames.end(), column) != fullFlattenedColumnNames.end();
+            return std::find(mainChunk.fullFlattenedColumnNames.begin(), mainChunk.fullFlattenedColumnNames.end(), column) != mainChunk.fullFlattenedColumnNames.end();
```
Comment (FinchPowers) [inline builtin/sub_dataset.cc]: Lines too long

File: builtin/sub_dataset.cc [hunk 4]
```diff
@@ -556,14 +660,30 @@ std::vector<ColumnPath>
 SubDataset::
 getFlattenedColumnNames() const
 {
-    return std::vector<ColumnPath>(itl->fullFlattenedColumnNames.begin(), itl->fullFlattenedColumnNames.end());
+    return std::vector<ColumnPath>(itl->mainChunk.fullFlattenedColumnNames.begin(), itl->mainChunk.fullFlattenedColumnNames.end());
```
Comment (FinchPowers) [inline builtin/sub_dataset.cc]: too long

File: builtin/sub_dataset.cc [hunk 5]
```diff
@@ -71,24 +76,102 @@ struct SubDataset::Itl
...
+
+        virtual void
+        recordRows(const std::vector<std::pair<RowPath, std::vector<std::tuple<ColumnPath, CellValue, Date> > > > & rows) override
+        {
+            for (auto & r: rows)
+                recordRow(r.first, r.second);
+        }
+
+        virtual void
+        recordRowsDestructive(std::vector<std::pair<RowPath, std::vector<std::tuple<ColumnPath, CellValue, Date> > > > rows) override
```
Comment (FinchPowers) [inline builtin/sub_dataset.cc]: Lines too long

**Comment by mathieumb** [comment]
@FinchPowers

**Pull request merged**